File: samples/run_five_node_100.py
#!/usr/bin/env python3
"""
五节点管线示例：模拟 100 帧输入，并在运行中动态 Configure 部分节点。

拓扑（5 节点）:
  app_src → resize → detector → rules → app_des

生命周期:
  1. Build(graph)
  2. Configure(初始 runtime)
  3. Start()
  4. PushFrame × 100（第 50 帧后热更新 resize / rules / 队列深度）
  5. Stop()
"""

from __future__ import annotations

import sys
import threading
import time
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from simple_pipe.context.frame_meta import FrameInput
from simple_pipe.pipeline.simple_pipeline import SimplePipeline
from simple_pipe.spec.runtime_config import PipelineRuntimeConfig

GRAPH = ROOT / "configs" / "demo" / "graph" / "five_node.json"
DEFAULTS_DIR = ROOT / "configs" / "demo" / "node_defaults"
FRAME_COUNT = 100
RUNTIME_PATCH_AT_FRAME = 50
# 推帧间隔（秒），避免瞬时打满队列导致大量 DROP_NEWEST 丢帧
PUSH_INTERVAL_S = 0.002


def main() -> None:
    pipeline = SimplePipeline(defaults_dir=DEFAULTS_DIR)

    build = pipeline.build_from_json(GRAPH)
    if not build.ok:
        raise RuntimeError(f"Build failed: {build.error}")

    # 初始运行时配置（Build 之后、Start 之前）
    initial_runtime = PipelineRuntimeConfig.from_dict(
        {
            "pipeline_params": {
                "app_src_in_queue_limit": 16,
                "app_des_in_queue_limit": 16,
            },
            "node_params": {
                "resize": {"width": 800, "height": 600},
                "rules": {"min_score": 0.5},
            },
        }
    )
    cfg = pipeline.configure(initial_runtime)
    if not cfg.ok:
        raise RuntimeError(f"Configure failed: {cfg.error}")

    received = 0
    alerts_before_patch = 0
    alerts_after_patch = 0
    resize_width_hist: Counter[int] = Counter()
    lock = threading.Lock()
    done = threading.Event()

    def on_output(meta) -> None:
        nonlocal received, alerts_before_patch, alerts_after_patch
        safety = meta.extensions.get("safety") or {}
        alert = bool(safety.get("alert"))
        resize_w = (meta.extensions.get("resize") or {}).get("width")
        with lock:
            received += 1
            if meta.frame_index < RUNTIME_PATCH_AT_FRAME:
                if alert:
                    alerts_before_patch += 1
            else:
                if alert:
                    alerts_after_patch += 1
            if resize_w is not None:
                resize_width_hist[resize_w] += 1
            if received >= FRAME_COUNT:
                done.set()

    pipeline.set_output_callback(on_output)
    if not pipeline.start().ok:
        raise RuntimeError("Start failed")

    print("Pipeline running — 5 nodes, pushing 100 frames …")
    print(f"  graph: {GRAPH.relative_to(ROOT)}")
    print(f"  runtime patch at frame index >= {RUNTIME_PATCH_AT_FRAME}")

    t0 = time.perf_counter()
    for i in range(FRAME_COUNT):
        if i == RUNTIME_PATCH_AT_FRAME:
            # 运行中热更新：改 resize 尺寸、提高 rules 阈值（节点级 patch）
            patch_runtime = PipelineRuntimeConfig.from_dict(
                {
                    "node_params": {
                        "resize": {"width": 1920, "height": 1080},
                        "rules": {"min_score": 0.95},
                    },
                }
            )
            mid = pipeline.configure(patch_runtime)
            if not mid.ok:
                pipeline.stop()
                raise RuntimeError(f"Mid-run Configure failed: {mid.error}")
            print(f"  [frame {i}] runtime Configure applied (resize=1920x1080, min_score=0.95)")

        result = pipeline.push_frame(
            FrameInput(
                image=f"mock-image-{i}",
                frame_index=i,
                channel_index=0,
            )
        )
        if not result.ok:
            pipeline.stop()
            raise RuntimeError(f"PushFrame failed at {i}: {result.error}")
        time.sleep(PUSH_INTERVAL_S)

    if not done.wait(timeout=30.0):
        pipeline.stop()
        raise TimeoutError(f"Only received {received}/{FRAME_COUNT} sink callbacks")

    elapsed = time.perf_counter() - t0
    pipeline.stop()

    print()
    print("=== Summary ===")
    print(f"  frames pushed:     {FRAME_COUNT}  (PushFrame × {FRAME_COUNT})")
    print(f"  sink callbacks:    {received}")
    print(f"  elapsed:           {elapsed:.3f}s ({FRAME_COUNT / elapsed:.1f} fps push)")
    print(f"  alerts [0, {RUNTIME_PATCH_AT_FRAME}):   {alerts_before_patch}")
    print(f"  alerts [{RUNTIME_PATCH_AT_FRAME}, 100): {alerts_after_patch}")
    print(f"  resize width hist: {dict(sorted(resize_width_hist.items()))}")
    print()
    print("Expected: before patch most frames alert (min_score=0.5);")
    print("          after patch fewer alerts (min_score=0.95); resize width shifts to 1920.")


if __name__ == "__main__":
    main()

File: samples/run_five_node_100.cpp
#include <atomic>
#include <chrono>
#include <iostream>
#include <thread>

#include "simple_pipe/context/frame_meta.hpp"
#include "simple_pipe/pipeline/simple_pipeline.hpp"
#include "simple_pipe/spec/runtime_config.hpp"

#ifndef SIMPLE_PIPE_SOURCE_DIR
#define SIMPLE_PIPE_SOURCE_DIR "."
#endif

int main() {
  const std::string root = SIMPLE_PIPE_SOURCE_DIR;
  simple_pipe::SimplePipeline pipeline(root + "/configs/demo/node_defaults");

  if (!(pipeline.BuildFromJson(root + "/configs/demo/graph/five_node.json")).IsSuccess()) {
    std::cerr << "Build failed\n";
    return 1;
  }

  nlohmann::json initial = {
      {"pipeline_params", {{"app_src_in_queue_limit", 16}, {"app_des_in_queue_limit", 16}}},
      {"node_params",
       {{"resize", {{"width", 800}, {"height", 600}}}, {"rules", {{"min_score", 0.5}}}}},
  };
  pipeline.Configure(simple_pipe::PipelineRuntimeConfig::FromJson(initial));

  constexpr int kFrames = 100;
  constexpr int kPatchAt = 50;

  std::atomic<int> received{0};
  std::atomic<int> alerts_before{0};
  std::atomic<int> alerts_after{0};

  pipeline.SetOutputCallback([&](const simple_pipe::FrameMeta& meta) {
    const int idx = meta.frame_index;
    const bool alert =
        meta.extensions.contains("safety") && meta.extensions["safety"].value("alert", false);
    if (idx < kPatchAt && alert) {
      alerts_before.fetch_add(1);
    } else if (idx >= kPatchAt && alert) {
      alerts_after.fetch_add(1);
    }
    received.fetch_add(1);
  });

  pipeline.StartPipeline();
  const auto t0 = std::chrono::steady_clock::now();

  for (int i = 0; i < kFrames; ++i) {
    if (i == kPatchAt) {
      nlohmann::json patch = {
          {"node_params",
           {{"resize", {{"width", 1920}, {"height", 1080}}},
            {"rules", {{"min_score", 0.95}}}}},
      };
      pipeline.Configure(simple_pipe::PipelineRuntimeConfig::FromJson(patch));
      std::cout << "[frame " << i << "] runtime Configure applied\n";
    }
    simple_pipe::FrameInput input;
    input.image = "mock-" + std::to_string(i);
    input.frame_index = i;
    if (!(pipeline.PushFrame(input)).IsSuccess()) {
      std::cerr << "PushFrame failed at " << i << '\n';
      return 1;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(2));
  }

  const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(30);
  while (received.load() < kFrames && std::chrono::steady_clock::now() < deadline) {
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  pipeline.StopPipeline();
  const auto elapsed = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();

  std::cout << "\n=== Summary ===\n"
            << "  frames pushed:   " << kFrames << '\n'
            << "  sink callbacks:  " << received.load() << '\n'
            << "  elapsed (s):     " << elapsed << '\n'
            << "  alerts [0," << kPatchAt << "):  " << alerts_before.load() << '\n'
            << "  alerts [" << kPatchAt << ",100): " << alerts_after.load() << '\n';

  return received.load() == kFrames ? 0 : 1;
}
